src/dtls13.c

```diff
@@ -0,0 +1,2187 @@
+    return method;
+}
+
+int Dtls13RlAddPlaintextHeader(
+    WOLFSSL *ssl, byte *out, enum ContentType content_type, size_t length)
+{
+    Dtls13RecordPlaintextHeader *hdr;
+    word32 seq[2];
+    int ret;
+
+    hdr = (Dtls13RecordPlaintextHeader *)out;
+    hdr->contentType = content_type;
+    hdr->legacyVersionRecord.major = DTLS_MAJOR;
+    hdr->legacyVersionRecord.minor = DTLSv1_2_MINOR;
+
+    ret = Dtls13GetSeq(ssl, CUR_ORDER, seq, 1);
+    if (ret != 0)
+        return ret;
+
+    /* seq[0] combines epoch and 16 MSB of sequence number */
+    c32toa(seq[0], hdr->epoch);
+    c32toa(seq[1], &hdr->sequenceNumber[2]);
+
+    c16toa(length, hdr->length);
+
+    return 0;
+}
+
+static int Dtls13HandshakeAddHeaderFrag(WOLFSSL *ssl, byte *output,
+    enum HandShakeType msg_type, word32 frag_offset, word32 frag_length,
+    word32 msg_length)
+{
+    Dtls13HandshakeHeader *hdr;
+
+    hdr = (Dtls13HandshakeHeader *)output;
+
+    hdr->msg_type = msg_type;
+    c32to24((word32)msg_length, hdr->length);
+    c16toa(ssl->keys.dtls_handshake_number, hdr->messageSeq);
+
+    c32to24(frag_offset, hdr->fragmentOffset);
+    c32to24(frag_length, hdr->fragmentLength);
+
+    return 0;
+}
+
+static int Dtls13TypeIsEncrypted(enum HandShakeType hs_type)
+{
+    switch (hs_type) {
+    case hello_request:
+    case hello_verify_request:
+    case client_hello:
+    case server_hello:
+        return 0;
+    case encrypted_extensions:
+    case session_ticket:
+    case end_of_early_data   :
+    case hello_retry_request :
+    case certificate         :
+    case server_key_exchange :
+    case certificate_request :
+    case server_hello_done   :
+    case certificate_verify  :
+    case client_key_exchange :
+    case finished            :
+    case certificate_status  :
+    case key_update          :
+    case change_cipher_hs    :
+    case message_hash        :
+    case no_shake            :
+        return 1;
+    }
+
+    /* abide compilers */
+    return 0;
+}
+
+static int Dtls13GetRnMask(
+    WOLFSSL *ssl, const byte *ciphertext, byte *mask, enum rnDirection dir)
+{
+    Ciphers *c;
+    int ret;
+
+    if (dir == PROTECT)
+        c = &ssl->dtlsRecordNumberEncrypt;
+    else
+        c = &ssl->dtlsRecordNumberDecrypt;
+
+    if (ssl->specs.bulk_cipher_algorithm == wolfssl_aes_gcm) {
+#ifdef HAVE_AESGCM
+        if (c->aes == NULL)
+            return BAD_STATE_E;
+        return  wc_AesEncryptDirect(c->aes, mask, ciphertext);
+#endif /* HAVE_AESGCM */
+    }
+
+    if (ssl->specs.bulk_cipher_algorithm == wolfssl_chacha) {
+#ifdef HAVE_CHACHA
+        word32 counter;
+
+        if (c->chacha == NULL)
+            return BAD_STATE_E;
+
+        /* assuming CIPHER[0..3] should be interpreted as big endian 32-bits
+           integer. The draft rfc isn't really clear on that. See sec 4.2.3 */
+        ato32(ciphertext, &counter);
+        ret = wc_Chacha_SetIV(c->chacha, &ciphertext[4], counter);
+        if (ret != 0)
+            return ret;
```

Comment:
I know we discussed this already but I will leave this comment to remember about this. I think that the byte stream should be interpreted as a little endian number per [Section 2.3 of the Chacha RFC](https://datatracker.ietf.org/doc/html/rfc7539#section-2.3).

Reply:
I still don't agree with this.

Reply:
Resolved to be LE.